### PR TITLE
Manage circular references in objects

### LINKED ISSUE #103: Problem with Parent-Leaf-Parent Relation in a watched object
Hi,

I identified a problem with the path generated by on-change when my object contains links to an other object, which contains itself a link to my original object...
For example, with something like this :

```
class Group {
  children: Layer[];
}
```

```
class Layer {
  parent: Group;
}
```

The found path will get longer and longer:

`group.children[0].children[0].parent.children[0].parent.children[0].parent.children[0].parent.children[0].parent.parent.children[0].parent.children[0].parent.children[0].children[0].parent.children[0].parent.children[0]...`

Actually, it seems that on-change falls in a infinite loop between children and parent.  
I've tried to configure the option `ignoreKeys`, but that actually not what I'm looking for, because I want to be notified when the parent or a child is modified.

I don't know if I am doing something wrong, but I think I should have an option to detect if a parent was already "verified for change" in a previous iteration, because `group.children[0].children[0].parent` is exactly the same object as `group.children[0]`, and the iteration should stop here. Is there a possibility to configure something like this ? 

I hope I'm clear enough in my description of the problem...

Many thanks!
Guillaume.

### PR DESCRIPTION
**Pull request by remyguillaume**
Hello,

I analyzed the problem described in the issue https://github.com/sindresorhus/on-change/issues/103 more in details, and I think this Pull-Request shoulb solve the problem.

The solution is to look first if the object already has a path within the same object, and if yes, not to generate a new path.

So actually if we are looking for the parth `group.children[0].children[0].parent` the object behind this path actually already exists with the path `group.children[0]`, and it's exactly the same object within the same tree structure. Therefore, the shortest path to the object should be used.

It's still possible to have the same object under many different path, this change in the code is only managing circular references within the same object.

The issue and the new code contain more comments and explanations.
Fixes #103

Could you please have a look at it and tell me if this change is ok for you?
Thanks !

And I forget the most important : Many thanks for your job, this library is really useful for us :)

### CODE REVIEW

**Comment by remyguillaume** [comment]
Ok I'll work on that, thanks.

I also just saw that there is an error with the linting, but I don't really understand why.
Could you please give me some explanations on what is wrong here?

**Comment by sindresorhus** [comment]
> Could you please give me some explanations on what is wrong here?

Fixed in `main`. Just rebase.

**Comment by remyguillaume** [comment]
I've added a test to reproduce the error and to validate the new code.

(Your changes on main are all visible in my pull-request now, I don't really understand why, but I read it seems to be a normal behavior in github.)

File: tests/on-change.test.js [hunk 1]
```diff
@@ -751,3 +751,35 @@ test('should not wrap a proxied object in another proxy', t => {
...
 	t.is(proxy.b, proxy.c);
 });
+
+test('path should be the shorter one in the same object for circular references', t => {
+	const layer1 = {group: null, val: 0};
+	const layer2 = {group: null, val: 0};
+	const layer3 = {group: null, val: 0};
+	const group = {
+		layers: [layer1, layer2, layer3],
+		val: 0,
```
Comment (sindresorhus) [inline tests/on-change.test.js:761]:
```suggestion
		value: 0,
```

File: tests/on-change.test.js [hunk 2]
```diff
@@ -751,3 +751,35 @@ test('should not wrap a proxied object in another proxy', t => {
...
+	};
+	layer1.group = group;
+	layer2.group = group;
+	layer3.group = group;
+
+	let resultPath = null;
```
Comment (sindresorhus) [inline tests/on-change.test.js:767]:
```suggestion
	let resultPath;
```

File: tests/on-change.test.js [hunk 3]
```diff
@@ -751,3 +751,35 @@ test('should not wrap a proxied object in another proxy', t => {
...
+	proxy.layers[0].group.val = 22;
+	t.is(resultPath, 'layers.0.group.val');
+
+	proxy.layers[0].group.layers[0].val = 33;
+	t.is(resultPath, 'layers.0.val');
+
+	proxy.layers[1].group.layers[0].group.layers[1].group.layers[1].group.layers[2].val = 33;
+	t.is(resultPath, 'layers.2.val');
+});
+
```
Comment (sindresorhus) [inline tests/on-change.test.js:785]: This test does not test the mentioned infinite loop. It would be good with a test that would have caused infinite loop before this pull request.

File: index.js [hunk 4]
```diff
@@ -74,7 +75,22 @@ const onChange = (object, onChange, options = {}) => {
...
+		// Check circular references
+		// If the value already has a corresponding path/proxy,
+		// And if the path corresponds to one of the parent of the current value
+		// Then we are on a circular case, where the child is pointing to his parent.
+		// => in this case we return the proxy object with the shortest path.
+		// Otherwise we can fell into an infinite loop,
+		// and the path can get longer and longer until we reach a memory limit.
+		const childPath = path.concat(basePath, property);
+		const existingPath = cache.getPath(value);
+		if (existingPath && !isArray(childPath) && !isSymbol(childPath) && childPath.toString().startsWith(existingPath.toString())) {
```
Comment (sindresorhus) [inline index.js:87]:
```suggestion
		if (existingPath && !isArray(childPath) && !isSymbol(childPath) && childPath.startsWith(existingPath.toString())) {
```

All tests pass without it. If it's actually needed, it should have a comment on what case it catches.

Comment (sindresorhus) [inline index.js:87]: Using `startsWith` is a bit fragile. It could potentially match the wrong thing. For example, `existingPath` is `foo` while `childPath` is `foo2.bar`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
> (Your changes on main are all visible in my pull-request now, I don't really understand why, but I read it seems to be a normal behavior in github.)

That's a sign of incorrect rebasing.

File: tests/on-change.test.js (hunk 3 above)
Comment (remyguillaume) [reply tests/on-change.test.js:785]: Sorry, actually "infinite loop" was not the right term here.  
The more we were modifying the object property, the more the path was getting longer.
This mean that after the first call of callback we got `proxy.layers[1].group.layers[0].val`.
After the second call we got `proxy.layers[1].group.layers[0].group.layers[2].group.layers[0].val`.
After the third call we got `proxy.layers[2].group.layers[2].group.layers[1].group.layers[0].group.layers[2].group.layers[0].val`.
... and so on.

Example fo logs:
image

And after 1 minute of usage of our app:
image

And actually we reached a memory limit because we were getting path that has Megabytes of data.

So, we cannot say this is a infinite loop.  
But this tests actually exactly covers our problem.

Comment (sindresorhus) [reply tests/on-change.test.js:785]: Ok. I would reword the comment then.

Comment (remyguillaume) [reply tests/on-change.test.js:785]: Ok, I've rewrited the comment.

File: index.js (hunk 4 above)
Comment (remyguillaume) [reply index.js]: Yes, you're totaly right, I didn't thought about this. I've rewriten the code to do a better comparison.
Actually I didn't manage the arrays of paths correctly in my first pull-request. I changed the code to manage arrays as well. And I've added another unittest for that too.

**Comment by remyguillaume** [comment]
> That's a sign of incorrect rebasing.
> 
> Your commit repeats: https://github.com/sindresorhus/on-change/pull/104/commits

Should I do something here ?
(Sorry, but I'm used to GitLab, which seems to manage Merge-Requests in a different way than GitHub does.)

**Comment by remyguillaume** [comment]
Hi @sindresorhus 

Thanks for merging, and have a merry Christmas ! 😊

**Comment by sindresorhus** [comment]
Merry Christmas :)
